JavaScript/Angular.md

```diff
@@ -0,0 +1,1551 @@
+
+So it is really up to you to decide whether you want to test some component with all its dependencies mocked or if you want to test interaction with other real components as well.
+
+What you call those Jasmine tests - unit or integration - doesn't really matter, ♪ *anyone can see* ♪.
+
+#### E2E testing
+
+End-to-end testing is quite a bit different when compared to unit/integration testing with Jasmine. Angular CLI projects come with a separate E2E testing project which uses Protractor. [Protractor](https://www.protractortest.org/) is basically a wrapper around [Selenium](https://www.seleniumhq.org/) which allows us to write tests in JavaScript and it also has some Angular-specific helper functions.
+
+When you start E2E tests, your application will be built and a programmatically controlled instance of a web browser will be opened and then the tests will click on various elements of the web-page without any human input (via [WebDriver](https://www.seleniumhq.org/projects/webdriver/)). This allows us to create quick smoke-testing type of tests which go through main flows of our application in an automated way.
+
+Covering E2E testing into detail is out of scope for this Handbook, please check out the [official documentation](https://angular.io/cli/e2e) if you would like to know more. However, we do have some quick tips and tricks:
+- `ng e2e` command builds your app and starts a local devserver just like `ng serve` and it then runs Protractor on that local instance of your app. If you want to run e2e tests on some specific environment instead of a local environment, you can run Protractor directly instead of via Angular CLI, simply by executing `protractor` (make sure to install it globally or use binary from node_modules or create a script in package.json). In `protractor.conf.js` you can configure `baseUrl` to point to the URL of your app on desired environment
+- You can use environment variables in Protractor tests, for example if you have some login functionality test, you can do something like: `$('input[data-e2e-test="login-email"]').sendKeys(process.env.E2E_LOGIN_EMAIL);`
+- You can set up some pre-conditions using `onPrepare` hook in `protractor.conf.js`
+- If you are using async/await in your e2e tests, make sure to [check this out](https://github.com/angular/protractor/blob/master/docs/async-await.md)
+
+One last thing we would like to mention is that Protractor isn't the only solution for e2e testing, but it does come out-of-the box with Angular CLI generated projects. Honestly, writing Protractor tests was often a painful process (especially if you are writing tests using [Cucumber](https://cucumber.io/)). There are some newer e2e testing frameworks which might be worth checking out:
+- [Cypress.io](https://www.cypress.io/)
+- [Puppeteer](https://developers.google.com/web/tools/puppeteer/)
```

Comment:
Then my list of alternatives has one thing xD I guess I can add vanilla selenium, do you know any other options? I think PhantomJS isn't even worth mentioning since it's dead. Maybe nightwatch?

Reply:
I think https://webdriver.io/ is a good alternative.

Reply:
Added webdriver but kept puppeteer because https://github.com/GoogleChrome/puppeteer#what-can-i-do `Automate form submission, UI testing, keyboard input, etc.` (added a note)